test/bio_dgram_test.c

```diff
@@ -0,0 +1,463 @@
+/*
+ * Copyright 2021 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/bio.h>
+#include "testutil.h"
+#include "internal/sockets.h"
+
+#if !defined(OPENSSL_NO_DGRAM) && !defined(OPENSSL_NO_SOCK)
+
+static int compare_addr(const BIO_ADDR *a, const BIO_ADDR *b)
+{
+    struct in_addr xa, xb;
+#if defined(OPENSSL_USE_IPV6)
+    struct in6_addr xa6, xb6;
+#endif
+    void *pa, *pb;
+    size_t slen, tmplen;
+
+    if (BIO_ADDR_family(a) != BIO_ADDR_family(b))
+        return 0;
+
+    if (BIO_ADDR_family(a) == AF_INET) {
+        pa = &xa;
+        pb = &xb;
+        slen = sizeof(xa);
+    }
+#if defined(OPENSSL_USE_IPV6)
+    else if (BIO_ADDR_family(a) == AF_INET6) {
+        pa = &xa6;
+        pb = &xb6;
+        slen = sizeof(xa6);
+    }
+#endif
+    else {
+        return 0;
+    }
+
+    tmplen = slen;
+    if (BIO_ADDR_rawaddress(a, pa, &tmplen) < 1)
```

Comment:
What's the reasoning for not using the TEST_ macros to generate a useful message?  Their use doesn't preclude the caller also generating a message to further pinpoint the location.